### PR TITLE
chore: document and cleanup decoder api

### PR DESCRIPTION
**Pull request by Alizter**
This is an attempt at documenting the undocumented parts of the decoder API and cleaning up unused values while we are at it.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Half the wrapping changes seem unnecessary and do not respect the 80 char rule.

**Comment by Alizter** [comment]
@rgrinberg I was not aware we had an 80 char limit, I didn't see one in hacking.rst. This is the char limit ocamlformat suggests and I made it consistent throughout the file.

**Comment by emillon** [comment]
I agree, these wrapping changes are a distraction. If that was just for the doc changes, this PR would have been merged already.

File: src/dune_sexp/decoder.mli [hunk 1]
```diff
@@ -194,51 +252,58 @@ val map_validate
...
-val leftover_fields_as_sums : (string * 'a t) list -> 'a list fields_parser
-
-(** Default value for [on_dup]. It fails with an appropriate error message. *)
-val field_present_too_many_times : Univ_map.t -> string -> Ast.t list -> _
-
+(** [leftover_fields] returns the remaining list of unparsed S-expressions. *)
 val leftover_fields : Ast.t list fields_parser
-val ( let* ) : ('a, 'k) parser -> ('a -> ('b, 'k) parser) -> ('b, 'k) parser
-val ( let+ ) : ('a, 'k) parser -> ('a -> 'b) -> ('b, 'k) parser
-val ( and+ ) : ('a, 'k) parser -> ('b, 'k) parser -> ('a * 'b, 'k) parser
```
Comment (Alizter) [inline src/dune_sexp/decoder.mli:244]: These have been moved up next to the other operators.

File: src/dune_sexp/decoder.mli [hunk 2]
```diff
@@ -2,59 +2,87 @@
...
+
+(** [( let* ) t f] is [t >>= fun x -> f x]. *)
+val ( let* ) : ('a, 'k) parser -> ('a -> ('b, 'k) parser) -> ('b, 'k) parser
+
+(** [( let+ ) t f] is [t >>| fun x -> f x]. *)
+val ( let+ ) : ('a, 'k) parser -> ('a -> 'b) -> ('b, 'k) parser
+
+(** [and+ t u] parses the input using [t] and then [u] and returns a pair of the two
+    results. This is the same as [pair]. *)
+val ( and+ ) : ('a, 'k) parser -> ('b, 'k) parser -> ('a * 'b, 'k) parser
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_sexp/decoder.mli:62]: Nope. `pair` is for parsing tuples formed as lists.

**Pull request merged**